sale_gathering/models/sale_order.py

```diff
@@ -0,0 +1,46 @@
+from odoo import models, fields, api, _
+from odoo.tools import float_compare
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+
+    _inherit = 'sale.order'
+
+    is_gathering = fields.Boolean('Is Gathering?')
+
+    amount_gathering = fields.Float(
+        compute="_compute_amount_gathering",
+        digits='Product Price',
+        )
+
+    @api.depends('order_line', 'is_gathering')
+    def _compute_amount_gathering(self):
+        self.amount_gathering = 0.0
+        for rec in self.filtered('is_gathering'):
+            amount_to_invoice = sum(
+                rec.order_line.filtered(lambda x: not x.is_downpayment).mapped(
+                    lambda l: l.price_unit_with_tax * l.qty_to_invoice))
+            amount_invoiced = sum(
+                rec.order_line.filtered(lambda x: not x.is_downpayment).mapped(
+                    lambda l: l.price_unit_with_tax * l.qty_invoiced))
+            rec.amount_gathering = sum(
+                rec.order_line.filtered('is_downpayment').mapped(
+                    'price_unit_with_tax')) - amount_invoiced - amount_to_invoice
+
+    def _get_invoiceable_lines(self, final=False):
+        """Return the invoiceable lines for order `self`."""
+        invoiceable_lines = super()._get_invoiceable_lines(final=False)
+        if self.is_gathering and self.amount_gathering > 0.0:
+            for line in self.order_line.filtered('is_downpayment'):
+                if final:
+                    invoiceable_lines |= line
+        return invoiceable_lines
+
+    @api.constrains('is_gathering', 'amount_total')
+    def _check_amount_gathering(self):
+        product_precision_digits = self.env['decimal.precision'].precision_get('Product Price')
+        for rec in self.filtered('is_gathering'):
+            if float_compare(rec.amount_gathering, 0.0, precision_digits=product_precision_digits) == -1:
+                raise ValidationError(_("The amount gathering will be negative, you cannot add more products"
```

Comment:
Dejaría así el mensaje

"The gathering balance will be negative (%s), you cannot make this modification to the order. Order: %s" % (rec.amount_gathering, rec.name)